pkg/deploy/strategy/support/lifecycle.go

```diff
@@ -70,29 +100,67 @@ func (e *HookExecutor) executeExecNewPod(hook *deployapi.LifecycleHook, deployme
 	}
 
 	// Try to create the pod.
-	pod, err := e.PodClient.CreatePod(deployment.Namespace, podSpec)
+	pod, err := e.podClient.CreatePod(deployment.Namespace, podSpec)
 	if err != nil {
 		if !kerrors.IsAlreadyExists(err) {
 			return fmt.Errorf("couldn't create lifecycle pod for %s: %v", deployutil.LabelForDeployment(deployment), err)
 		}
 	} else {
-		glog.V(0).Infof("Created lifecycle pod %s for deployment %s", pod.Name, deployutil.LabelForDeployment(deployment))
+		glog.V(0).Infof("Created lifecycle pod %s/%s for deployment %s", pod.Namespace, pod.Name, deployutil.LabelForDeployment(deployment))
 	}
 
 	stopChannel := make(chan struct{})
 	defer close(stopChannel)
-	nextPod := e.PodClient.PodWatch(pod.Namespace, pod.Name, pod.ResourceVersion, stopChannel)
-
-	glog.V(0).Infof("Waiting for hook pod %s/%s to complete", pod.Namespace, pod.Name)
+	nextPod := e.podClient.PodWatch(pod.Namespace, pod.Name, pod.ResourceVersion, stopChannel)
+
+	// Wait for the hook pod to reach a terminal phase. Start reading logs as
+	// soon as the pod enters a usable phase.
+	var updatedPod *kapi.Pod
+	var once sync.Once
+	wg := &sync.WaitGroup{}
+	wg.Add(1)
+	glog.V(0).Infof("Watching logs for hook pod %s/%s while awaiting completion", pod.Namespace, pod.Name)
+waitLoop:
 	for {
-		pod := nextPod()
-		switch pod.Status.Phase {
-		case kapi.PodSucceeded:
-			return nil
-		case kapi.PodFailed:
-			return fmt.Errorf(pod.Status.Message)
+		updatedPod = nextPod()
+		switch updatedPod.Status.Phase {
+		case kapi.PodRunning:
+			go once.Do(func() { e.readPodLogs(pod, wg) })
+		case kapi.PodSucceeded, kapi.PodFailed:
+			go once.Do(func() { e.readPodLogs(pod, wg) })
+			break waitLoop
 		}
 	}
+	// The pod is finished, wait for all logs to be consumed before returning.
+	wg.Wait()
+	if updatedPod.Status.Phase == kapi.PodFailed {
+		return fmt.Errorf(updatedPod.Status.Message)
+	}
+	return nil
+}
+
+// readPodLogs streams logs from pod to podLogDestination. It signals wg when
+// done.
+func (e *HookExecutor) readPodLogs(pod *kapi.Pod, wg *sync.WaitGroup) {
+	defer wg.Done()
+	opts := &kapi.PodLogOptions{
+		Container:  HookContainerName,
+		Follow:     true,
+		Timestamps: true,
+	}
+	logStream, err := e.podLogStream(pod.Namespace, pod.Name, opts)
+	if err != nil || logStream == nil {
+		glog.V(0).Infof("Warning: couldn't get log stream for lifecycle pod %s/%s: %s", pod.Namespace, pod.Name, err)
+		return
+	}
+	// Read logs.
+	defer logStream.Close()
+	written, err := io.Copy(e.podLogDestination, logStream)
+	if err != nil {
+		glog.V(0).Infof("Finished reading logs for hook pod %s/%s (%d bytes): %s", pod.Namespace, pod.Name, written, err)
+	} else {
+		glog.V(0).Infof("Finished reading logs for hook pod %s/%s", pod.Namespace, pod.Name)
```

Comment:
I was sending them for the end user. There are clear lines of demarcation in the logs indicating when the hook pods begin/end.